keanu-project/src/main/java/io/improbable/keanu/network/ModelComposition.java

```diff
@@ -0,0 +1,98 @@
+     */
+    public static Map<VertexLabel, Vertex> createModelVertices(BayesianNetwork bayesianNetwork,
+                                                               Map<VertexLabel, Vertex> inputVertices,
+                                                               List<VertexLabel> desiredOutputs) {
+        Map<VertexLabel, Vertex> outputMap = extractOutputs(bayesianNetwork, desiredOutputs);
+        increaseDepth(bayesianNetwork, outputMap);
+        checkAndLinkInputs(bayesianNetwork, inputVertices);
+        cleanOutputLabels(outputMap);
+
+        return outputMap;
+    }
+
+    private static Map<VertexLabel, Vertex> extractOutputs(BayesianNetwork bayesianNetwork,
+                                                           List<VertexLabel> desiredOutputs) {
+        if (desiredOutputs.size() == 0) {
+            throw new IllegalArgumentException("At least one output must be specified");
+        }
+
+        Map<VertexLabel, Vertex> outputMap = new HashMap<>();
+
+        for (VertexLabel label: desiredOutputs) {
+            Vertex v = bayesianNetwork.getVertexByLabel(label);
+            if (v == null) {
+                throw new IllegalArgumentException("Unable to find Output Vertex: " + label);
+            }
+            outputMap.put(label, v);
+        }
+
+        return outputMap;
+    }
+
+    private static void increaseDepth(BayesianNetwork bayesianNetwork, Map<VertexLabel, Vertex> outputVertices) {
+        VertexId newPrefix = new VertexId();
+        bayesianNetwork.incrementDepth();
+        bayesianNetwork.getAllVertices().stream()
+            .filter(v -> !outputVertices.containsKey(v.getLabel()))
+            .forEach(v -> v.getId().increaseDepth(newPrefix));
+        bayesianNetwork.getAllVertices().stream()
+            .filter(v -> outputVertices.containsKey(v.getLabel()))
+            .forEach(v -> v.getId().resetID());
+    }
+
+    private static void checkAndLinkInputs(BayesianNetwork bayesianNetwork, Map<VertexLabel, Vertex> inputs) {
+        for (Map.Entry<VertexLabel, Vertex> entry: inputs.entrySet()) {
+            Vertex v = bayesianNetwork.getVertexByLabel(entry.getKey());
+
+            if (v == null) {
+                throw new IllegalArgumentException("No node labelled \"" + entry.getKey() + "\" found");
+            }
+
+            if (v instanceof ProxyVertex) {
+                ProxyVertex proxyVertex = (ProxyVertex)v;
+                if (proxyVertex.hasParent()) {
+                    throw new IllegalArgumentException("Proxy Vertex for \"" + v.getLabel() + "\" already has Parent");
+                } else {
+                    proxyVertex.setParent(entry.getValue());
+                }
+            } else {
+                throw new IllegalArgumentException("Input node \"" + entry.getKey() + "\" is not a Proxy Vertex");
+            }
+        }
+    }
+
+    private static void cleanOutputLabels(Map<VertexLabel, Vertex> outputMap) {
```

Comment:
Why don't we clean input and proxy vertices?

Reply:
This is a little subtle and relates to how I've ensured that we create a "namespace" for labels.
What I'm trying to avoid is a situation, for instance, where we use two instances of a subgraph - both of these will have Vertices with the same labels - meaning we have a clash of labels in the outer graph.

The output vertices themselves "pop out" of the inner graph and exist at the same level as the parent graph whereas all the other vertices get pushed down to a lower depth.  When creating the label map in the BayesNet I only look at vertices at the top level - effectively creating separate "namespaces" for Vertex Labels.  The exception is obviously the output vertices which will be picked up, so I clean the labels explicitly for this.

This is all sort of temporary stuff until we have the full functionality to "instantiate" graphs from models at which point a lot of this stuff can be cleaned up a bit.

Reply:
Instead of having the output vertices "pop out" of the inner graph, would it be possible (and not too difficult) to leave them (labelled) in the inner graph but link them to (new) vertices in the outer graph? That way it's consistent with how you link the input vertices, and probably easier to understand how all this works.

Reply:
I'll chat to you Mike - it's not ideal to not pop them out (ie I can't use the same proxy vertex trick I've used for inputs) and this is the most elegant solution I could think of.